src/search_controller.ts

```diff
@@ -0,0 +1,484 @@
+  readonly isActive: boolean;
+  readonly isLoading: boolean;
+  readonly items: T[] | undefined;
+  readonly lastQueryError: Error | undefined;
+  readonly next: string | undefined;
+  readonly offset: number | undefined;
+  resetState(stateOverrides?: Partial<SearchSourceState<T>>): void;
+  search(text?: string): void;
+  searchDebounced: DebouncedExecQueryFunction;
+  readonly searchQuery: string;
+  setDebounceOptions(options: DebounceOptions): void;
+  readonly state: StateStore<SearchSourceState<T>>;
+  readonly type: SearchSourceType;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type SearchSourceState<T = any> = {
+  hasMore: boolean;
+  isActive: boolean;
+  isLoading: boolean;
+  items: T[] | undefined;
+  searchQuery: string;
+  lastQueryError?: Error;
+  next?: string;
+  offset?: number;
+};
+
+export type SearchSourceOptions = {
+  /** The number of milliseconds to debounce the search query. The default interval is 300ms. */
+  debounceMs?: number;
+  isActive?: boolean;
+  pageSize?: number;
+};
+
+const DEFAULT_SEARCH_SOURCE_OPTIONS: Required<SearchSourceOptions> = {
+  debounceMs: 300,
+  isActive: false,
+  pageSize: 10,
+} as const;
+
+export abstract class BaseSearchSource<T> implements SearchSource<T> {
+  state: StateStore<SearchSourceState<T>>;
+  protected pageSize: number;
+  abstract readonly type: SearchSourceType;
+  searchDebounced!: DebouncedExecQueryFunction;
+
+  protected constructor(options?: SearchSourceOptions) {
+    const { debounceMs, isActive, pageSize } = { ...DEFAULT_SEARCH_SOURCE_OPTIONS, ...options };
+    this.pageSize = pageSize;
+    this.state = new StateStore<SearchSourceState<T>>({
+      hasMore: true,
+      isActive,
+      isLoading: false,
+      items: undefined,
+      offset: 0,
+      searchQuery: '',
+    });
+    this.setDebounceOptions({ debounceMs });
+  }
+
+  get lastQueryError() {
+    return this.state.getLatestValue().lastQueryError;
+  }
+
+  get hasMore() {
+    return this.state.getLatestValue().hasMore;
+  }
+
+  get hasResults() {
+    return Array.isArray(this.state.getLatestValue().items);
+  }
+
+  get isActive() {
+    return this.state.getLatestValue().isActive;
+  }
+
+  get isLoading() {
+    return this.state.getLatestValue().isLoading;
+  }
+
+  get items() {
+    return this.state.getLatestValue().items;
+  }
+
+  get next() {
+    return this.state.getLatestValue().next;
+  }
+
+  get offset() {
+    return this.state.getLatestValue().offset;
+  }
+
+  get searchQuery() {
+    return this.state.getLatestValue().searchQuery;
+  }
+
+  protected abstract query(searchQuery: string): Promise<QueryReturnValue<T>>;
+
+  protected abstract filterQueryResults(items: T[]): T[] | Promise<T[]>;
+
+  setDebounceOptions = ({ debounceMs }: DebounceOptions) => {
+    this.searchDebounced = debounce(this.executeQuery.bind(this), debounceMs);
+  };
+
+  activate = (sourceStateOverride?: Partial<SearchSourceState>) => {
```

Comment:
Why the partial state here? I'd say `activate` should flip `active` to `true` and all the other side-effects should happen elsewhere. Multiple `StateStore.next` calls seem better than unexpected state changes seemingly unrelated to the method call.

Reply:
Removed partial state update on activation: https://github.com/GetStream/stream-chat-js/pull/1434/commits/072930d0afcba044b8d854b0fdb1ddc8e35c0d32